File: backend/app/models/followup.py
from sqlalchemy import String, Text, Float, DateTime, Boolean, ForeignKey, Integer
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func
from app.database import Base
import uuid


class LeadActivity(Base):
    """Registro de cualquier interacción/evento sobre un lead."""
    __tablename__ = "lead_activities"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    business_id: Mapped[str] = mapped_column(String, ForeignKey("businesses.id"), nullable=False, index=True)
    lead_id: Mapped[str] = mapped_column(String, ForeignKey("leads.id"), nullable=False, index=True)

    # Tipo: call | email | whatsapp | meeting | note | stage_change | followup_completed | ai_action
    activity_type: Mapped[str] = mapped_column(String(50), nullable=False, default="note")
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=True)

    # Resultado de la actividad: contacted | no_answer | interested | not_interested | scheduled | other
    outcome: Mapped[str] = mapped_column(String(50), nullable=True)

    # Si fue una actividad programada, cuándo estaba planificada y cuándo se ejecutó
    scheduled_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=True)

    # Quién la creó: "ai" o user_id
    created_by: Mapped[str] = mapped_column(String(100), nullable=False, default="manual")
    is_ai_generated: Mapped[bool] = mapped_column(Boolean, default=False)

    created_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    # Relationships
    lead: Mapped["Lead"] = relationship("Lead", back_populates="activities")  # type: ignore


class FollowUp(Base):
    """Seguimiento programado sobre un lead, con notificaciones y gestión IA/manual."""
    __tablename__ = "followups"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    business_id: Mapped[str] = mapped_column(String, ForeignKey("businesses.id"), nullable=False, index=True)
    lead_id: Mapped[str] = mapped_column(String, ForeignKey("leads.id"), nullable=False, index=True)

    # Tipo: call | email | whatsapp | meeting | task
    followup_type: Mapped[str] = mapped_column(String(50), nullable=False, default="call")

    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=True)

    # Estado: pending | completed | cancelled | overdue
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="pending", index=True)

    # Prioridad: low | medium | high | urgent
    priority: Mapped[str] = mapped_column(String(20), nullable=False, default="medium")

    scheduled_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=False, index=True)
    completed_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=True)

    # Asignado a: "ai" o user_id
    assigned_to: Mapped[str] = mapped_column(String(100), nullable=False, default="ai")
    is_ai_generated: Mapped[bool] = mapped_column(Boolean, default=False)

    # Notificaciones
    notify_email: Mapped[bool] = mapped_column(Boolean, default=True)
    notify_whatsapp: Mapped[bool] = mapped_column(Boolean, default=False)
    notification_sent_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), nullable=True)

    # Cuántas veces se ha recordado (para evitar spam)
    reminder_count: Mapped[int] = mapped_column(Integer, default=0)

    created_by: Mapped[str] = mapped_column(String(100), nullable=False, default="manual")

    created_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), onupdate=func.now(), nullable=True)

    # Relationships
    lead: Mapped["Lead"] = relationship("Lead", back_populates="followups")  # type: ignore

File: frontend/src/types/followup.ts
export interface FollowUp {
  id: string;
  lead_id: string;
  lead_name?: string;
  lead_email?: string;
  lead_company?: string;
  lead_stage?: string;
  followup_type: "call" | "email" | "whatsapp" | "meeting" | "task";
  title: string;
  description?: string;
  status: "pending" | "completed" | "cancelled" | "overdue";
  priority: "low" | "medium" | "high" | "urgent";
  scheduled_at: string;
  completed_at?: string;
  assigned_to: string;
  is_ai_generated: boolean;
  notify_email: boolean;
  notify_whatsapp: boolean;
  created_by: string;
  created_at: string;
  updated_at?: string;
}

export interface LeadActivity {
  id: string;
  lead_id: string;
  activity_type: string;
  title: string;
  description?: string;
  outcome?: string;
  scheduled_at?: string;
  completed_at?: string;
  created_by: string;
  is_ai_generated: boolean;
  created_at: string;
}

export interface FollowUpStats {
  overdue: number;
  today: number;
  this_week: number;
  total_pending: number;
  completed_today: number;
}

export const FOLLOWUP_TYPES = [
  { value: "call",     label: "Llamada",    icon: "📞", color: "text-blue-600 bg-blue-50",   gradient: "linear-gradient(135deg,#3B82F6,#6366F1)" },
  { value: "email",    label: "Email",      icon: "📧", color: "text-violet-600 bg-violet-50", gradient: "linear-gradient(135deg,#635BFF,#8B5CF6)" },
  { value: "whatsapp", label: "WhatsApp",   icon: "💬", color: "text-green-600 bg-green-50",  gradient: "linear-gradient(135deg,#10B981,#059669)" },
  { value: "meeting",  label: "Reunión",    icon: "🤝", color: "text-orange-600 bg-orange-50", gradient: "linear-gradient(135deg,#F97316,#EA580C)" },
  { value: "task",     label: "Tarea",      icon: "✅", color: "text-slate-600 bg-slate-50",  gradient: "linear-gradient(135deg,#64748B,#475569)" },
] as const;

export const PRIORITY_CONFIG = {
  low:    { label: "Baja",    color: "text-slate-500  bg-slate-50  border-slate-200" },
  medium: { label: "Media",   color: "text-blue-600   bg-blue-50   border-blue-200" },
  high:   { label: "Alta",    color: "text-orange-600 bg-orange-50 border-orange-200" },
  urgent: { label: "Urgente", color: "text-red-600    bg-red-50    border-red-200" },
} as const;

export const STATUS_CONFIG = {
  pending:   { label: "Pendiente",   color: "text-blue-600   bg-blue-50   border-blue-200" },
  overdue:   { label: "Vencido",     color: "text-red-600    bg-red-50    border-red-200" },
  completed: { label: "Completado",  color: "text-green-600  bg-green-50  border-green-200" },
  cancelled: { label: "Cancelado",   color: "text-slate-400  bg-slate-50  border-slate-200" },
} as const;
